pkg/storage/multi_sst_writer.go

```diff
@@ -0,0 +1,250 @@
+// Copyright 2023 The LevelDB-Go and Pebble Authors. All rights reserved. Use
+// of this source code is governed by a BSD-style license that can be found in
+// the LICENSE file.
+
+package storage
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/pebble"
+	"github.com/cockroachdb/pebble/internal/keyspan"
+	"github.com/cockroachdb/pebble/rangekey"
+	"github.com/cockroachdb/pebble/sstable"
+)
+
+// MultiSSTWriter may be used to construct a series of non-overlapping sstables
+// from a stream of keys. The user may specify a set of fixed boundaries such
+// that no sstable produced will span the boundaries.
+type MultiSSTWriter struct {
+	// It is up to the caller to provide an SSTWriter to be used for creating a new file.
+	getNewSSTFile func(ctx context.Context) (SSTWriter, error)
+	// uponFinish is called once an
+	beforeSSTFinalize func(start, end roachpb.Key) error
+	currSST           SSTWriter
+	keySpans          []roachpb.Span
+	currSpan          int
+	// The total size of SST data. Updated on SST finalization.
+	dataSize      int64
+	tableMetadata []*sstable.WriterMetadata
+	// Used for fragmenting range keys.
+	rangeKeyFrag keyspan.Fragmenter
+	// Used for fragmenting range deletes.
+	rangeDelFrag keyspan.Fragmenter
```

Comment:
`sstable.Writer` also has a `fragmenter` (maybe it serves a similar purpose) but I was not too sure